### ISSUE TITLE
Early stopping criteria is only checked for the `ReduceLROnPlateau` scheduler

### ISSUE THREAD

**Issue by mcognetta** [joeynmt/joeynmt#215]
**Describe the bug**
Early stopping and patience is only considered in the `ReduceLROnPlateau`. For example, using `bleu` as an early stopping criteria and the `WarmupInverseSquareRootScheduler` will not stop training early, even if validation performance does not improve for more than `patience` iterations.

The issue is in these lines: https://github.com/joeynmt/joeynmt/blob/0968187ac0968007cabebed5e5cb6587c08dff78/joeynmt/training.py#L532-L540

There should be an additional `stats` variable checking if validation performance has not increase in `N` iterations.

Side note: Sorry for the influx of issues (there will be a few more in coming days). I am getting ramped up on this project, so I have been doing lots of exploring. I will help to contribute fixes to anything that I report here.

**To Reproduce**
1. Use any of the basic training configurations.
2. Change the `scheduling` parameter to `scheduling: "warmupinversesquareroot"`
3. Run the training run.
4. Observe that training does not stop until max epochs, even if the validation performance stops improving.

**Logged output**
If possible, please add the training log and validation reports or excerpts of them.

**Expected behavior**
Training stops early when validation stops improving.

**System (please complete the following information):**
 - OS: Linux
 - CPU & GPU
 - Python3.10

**Comment by may-** [comment]
@mcognetta 
Thank you for reporting!
I think it's not a bug, but a feature request.

Actually, we initially introduced the `patience` parameter for the plateau scheduler only.
https://github.com/joeynmt/joeynmt/blob/0968187ac0968007cabebed5e5cb6587c08dff78/configs/transformer_small.yaml#L98

I'd say, the term `patience` in plateau scheduler won't directly control the early-stopping, but the LR reduction. Early-stopping is triggered when the LR reached the specified minimum value, in case of plateau scheduler.

Joeynmt doesn't stop the iteration, but skip to save the weights, if the validation performance has not increased.
https://github.com/joeynmt/joeynmt/blob/0968187ac0968007cabebed5e5cb6587c08dff78/joeynmt/training.py#L654-L658

I'm not sure how many users wish to stop the iteration instead of just skipping the checkpoint. I'd like to discuss further!!

**Comment by may-** [comment]
I understand there is such a situation and would be nice.

My concern is how essential the functionality is, though. You know, I also implement many small "nice-to-have"s, but keep them in my personal fork and don't merge them to the main branch.
To me, this sounds more like a job of a hyperparam tuning tool such as [optuna](https://optuna.org/) (they have pruning callbacks) and not necessarily coded in official joey. But yeah, it's just my preference, maybe. If joey users are happy with it, I have of course no objection at all 😄 

Thank you again for your effort and deep consideration! I really appreciate that. I'm so glad to see our joey is still growing 🐨

**Comment by mcognetta** [comment]
Understood, ofc I will defer to your judgement. However, do you mind if I open (and then immediately close) a PR here linked to this issue so that others can use it in the future if they need it (adding the relevant code snippets here is a little cumbersome and it would be good for posterity)? It may also serve as a place for the discussion about if this feature is wanted by many users or not.

Thanks!

**Comment by may-** [comment]
Yes, open a PR, please! I'm more than happy for democratic decision.
Honestly, I'm sometimes afraid that this project would have some bias towards my personal preferences. Always nice to have fresh ideas and manage with many eyes. Thank you for your effort!!
